File: src/core/execution/execution_context.rs
use crate::{decoding::DecodingContext, execution::processor::ProcessorFault};

use super::machine::ItaniumMachine;

use colored::*;

pub struct ExecutionContext<'a, 'b> {
    pub decoding_context: DecodingContext<'a>,
    pub machine: &'b mut ItaniumMachine,
    pub instruction_index: usize,
    pub paused: bool
}

impl ExecutionContext<'_, '_> {
    pub fn new<'a, 'b>(decoding_context: DecodingContext<'a>, machine: &'b mut ItaniumMachine) -> ExecutionContext<'a, 'b> {
        ExecutionContext { 
            decoding_context: decoding_context, 
            machine: machine, 
            instruction_index: 0, 
            paused: false 
        }
    }

    pub fn step(&mut self) {
        let executable = &self.decoding_context.executable_instructions[self.instruction_index];
        
        self.instruction_index += 1;

        let execution_result = (executable.execution_function)(&mut self.machine, &executable.attributes, &mut self.instruction_index, &self.decoding_context.instruction_index_to_address, &self.decoding_context.address_to_instruction_index);
    }

    pub fn pause(&mut self) {
        self.paused = true
    }
    
    pub fn run(&mut self) {
        while self.machine.continue_running && !self.paused {
            self.step()
        }
    }
}

File: Rosalia.Core/Execution/ExecutionContext.cs
using Rosalia.Core.Decoding;

namespace Rosalia.Core.Execution;

public delegate ProcessorFault ExecutionFunction(ExecutionContext context, Dictionary<InstructionAttribute, ulong> attributes);

public struct ExecutableInstruction {
    public ExecutionFunction                       ExecutionFunction;
    public Dictionary<InstructionAttribute, ulong> Attributes;
    public string                                  Disassembly;

    public override string ToString() => this.Disassembly;
}

public class ExecutionContext {
    public readonly DecodingContext DecodingContext;
    public readonly ItaniumMachine  Machine;

    private bool            _paused;

    public ExecutionContext(DecodingContext decodingContext, ItaniumMachine machine) {
        this.DecodingContext = decodingContext;
        this.Machine         = machine;

        this._paused          = false;
    }

    public void Step() {
        ulong instructionIndex = this.DecodingContext.ConvertInstructionPointer(this.Machine.Processor.InstructionPointer);
        ExecutableInstruction instruction = this.DecodingContext.ExecutableInstructions[(int)instructionIndex];

        ProcessorFault fault = instruction.ExecutionFunction(this, instruction.Attributes);

        if (fault != ProcessorFault.None) {
            Console.WriteLine($"Fault thrown: {fault}; On: {instruction.Disassembly}");
        }
    }

    public void Pause() {
        this._paused = true;
    }

    public void Run() {
        while (this.Machine.ContinueRunning && !this._paused) {
            this.Step();
        }

        Console.WriteLine($"IA64 Return Code: {this.Machine.Processor.GeneralRegisters.RetrieveRegister(8).RetrieveValue()}");
    }
}
